### DISCUSSION TITLE
Python 🐍 Programming Language Thread

### DISCUSSION THREAD

**Discussion by queenofcorgis** [community/community#51390] [Programming Help]
[Python](https://github.com/python/cpython) is a programming language that lets you work quickly and efficiently. _$ denotes a paid resource._ 

**Around the web**

- [Python Programming Exercises](https://github.com/zhiwehu/Python-programming-exercises) 
- [Python for Programmers Tutorials](https://wiki.python.org/moin/BeginnersGuide/Programmers)
- [Learn Python](https://github.com/trekhleb/learn-python) 
- [Intro to Python](https://youtu.be/WIFCd1m7VuA) 
- $ [Learn Python the Hard Way](http://learnpythonthehardway.org/) 

**Books**

- The Art of Clean Code: Best Practices to Eliminate Complexity and Simplify Your Life - Christian Mayer 
- Learn to Code by Solving Problems: A Python Programming Primer - Daniel Zingaro
- The Python Coding Book: Understanding what programming really is - Steph Grupetta 

Share any links, tutorials, videos, and books that have helped you learn Python below - don’t forget to include if it is paid too!

**Comment by sergeyleschev** [comment]
- Official Python documentation: https://docs.python.org/3/
- Python for Everybody course by Dr. Charles Severance: https://www.py4e.com/
- Automate the Boring Stuff with Python by Al Sweigart: https://automatetheboringstuff.com/
- Python Crash Course by Eric Matthes: https://nostarch.com/pythoncrashcourse2e
- Codecademy's Python course: https://www.codecademy.com/learn/learn-python

**Reply by queenofcorgis** [reply]
Very nice - appreciate you commenting with these @sergeyleschev

**Comment by APolevanov** [comment]
Around the web:

    Real Python: https://realpython.com/ - Offers a mix of free and paid tutorials, articles, and courses on various Python topics.
    Corey Schafer's Python Tutorials: https://www.youtube.com/playlist?list=PL-osiE80TeTt2d9bfVyTiXJA-UTHn6WwU - A YouTube playlist covering various Python concepts.
    Python.org's Beginner's Guide: https://docs.python.org/3/tutorial/index.html - The official Python tutorial for beginners.
    Edabit Python Challenges: https://edabit.com/challenges/python3 - Python coding challenges with varying difficulty levels.
    $ Codecademy's Python Course: https://www.codecademy.com/learn/learn-python - An interactive Python course for beginners.

Books:

    Python Crash Course: A Hands-On, Project-Based Introduction to Programming - Eric Matthes
    Automate the Boring Stuff with Python: Practical Programming for Total Beginners - Al Sweigart
    Fluent Python: Clear, Concise, and Effective Programming - Luciano Ramalho
    $ Effective Python: 90 Specific Ways to Write Better Python - Brett Slatkin
    $ Python Tricks: A Buffet of Awesome Python Features - Dan Bader

**Reply by queenofcorgis** [reply]
Wow - thanks for sharing all of these @APolevanov 🐍

**Comment by d-coder111** [comment]
I would like to share this amazing free CS50's Introduction to Programming with Python: https://www.edx.org/course/cs50s-introduction-to-programming-with-python?utm_source=lms_catalog_service_user&utm_medium=affiliate_partner

Hope it helps 🚀😍.

**Reply by Werunga91** [reply]
Keenly following this one a good Course with every updates.CS50P

**Comment by stoliarSant** [comment]
from tkinter import *
import random
import string, random


# codigo de combinações de formulas/ A IDEIA FOI CRIADA COM BASE NA COLETA DE COMBINAÇÕES DE NUMEROS('FORMULAS') GERANDO RESULTADOS ALEATORIOS("AUMENTOU A DIVERSIDADE DE RESULTADOS")

def movendo_digitos ():
    while True:
        BAGAGEM = [0,0,0]

            
            
        campo = ["PARTIDA",int(32),int(66),int(99),"CHEGADA"]
        
        regras_do_jogo = [int(14),int(8),int(9),int(5),int(17),int(16),int(24),int(16),int(33),int(34),int(33),int(334),int(133),int(343),int(733),int(933),int(3333)]
        
        personagem = float(0.1)#FUNÇÃO EM ANDAMENTO
        tiro = random.choice(regras_do_jogo)
        tiro_1 = random.choice(regras_do_jogo)

        conta = int(tiro+tiro_1)
        print(conta)
        if conta==campo[1]:
            BAGAGEM.insert(0,conta)
            print('PRIMEIRA PORTA')
            print(conta)
            #BAGAGEM.insert(0,conta)
        tiro_2 = random.choice(regras_do_jogo)
        if BAGAGEM[0]+tiro_2 == campo[2]:
            print("REGISTRADO = ",BAGAGEM[0]+tiro_2)
            conta_0 = BAGAGEM[0]+tiro_2
            BAGAGEM.insert(1,conta_0)
            print(tiro_2)
            
        tiro_3 = random.choice(regras_do_jogo)
        if BAGAGEM[1]+tiro_3 == campo[3]:
            print("REGISTRADO TERCEIRA PORTA= ",BAGAGEM[1]+tiro_3)
            conta_1 = BAGAGEM[1]+tiro_3
            BAGAGEM.insert(2,conta_1)
            print(tiro_3)
            print(BAGAGEM)
            break

       # input('tr')




movendo_digitos ()

**Comment by uvpatel** [comment]
## 🐍 Python Learning Resources

Python is a powerful programming language that helps you build things quickly and efficiently.  
> 💲 `$` denotes a paid resource.

---

## 🌐 Around the Web

- 🔗 Python Programming Exercises  
- 🔗 Python for Programmers Tutorials  
- 🔗 Learn Python  
- 🔗 Intro to Python  
- 💲 Learn Python the Hard Way  

---

## 📚 Books

- 📖 *The Art of Clean Code: Best Practices to Eliminate Complexity and Simplify Your Life* — Christian Mayer  
- 📖 *Learn to Code by Solving Problems: A Python Programming Primer* — Daniel Zingaro  
- 📖 *The Python Coding Book: Understanding What Programming Really Is* — Steph Grupetta  

---

## 🎥 Video Resources

- ▶️ Python Full Course (Beginner to Advanced)  
- ▶️ Python Projects for Practice  
- ▶️ Data Science & Machine Learning with Python  

---

## 🚀 Practice Platforms

- 💻 LeetCode (Python DSA Practice)  
- 💻 HackerRank (Beginner Friendly)  
- 💻 Codeforces (Competitive Programming)  

---

## 🙌 Contribute

Share any links, tutorials, videos, or books that helped you learn Python!  
Don’t forget to mention if the resource is **free or paid ($)**.

**Comment by buihuutam05062002** [comment]
https://www.w3schools.com/
